### PR TITLE
Number proxies

### PR DESCRIPTION
**Pull request by jjsjann123**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes # (issue).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: thunder/core/jit_ext.py [hunk 1]
```diff
@@ -595,6 +595,9 @@ def proxify(self, value: WrappedValue) -> Any:
             co: CACHE_OPTIONS = get_cache_option()
             if co is CACHE_OPTIONS.CONSTANT_VALUES:
                 self.add_constraint((clang.check_tensor_shape_and_metadata, p_orig))
+            elif co is CACHE_OPTIONS.SYMBOLIC_VALUES:
```
Comment (t-vi) [inline thunder/core/jit_ext.py:598]: What is our medium term plan w.r.t. defaul caching? If we need this for correctly handling #231 , it would seem that symbolic values should be the default but that in turn would mean that we want to have it work for our supported use cases.

Comment (jjsjann123) [reply thunder/core/jit_ext.py:598]: I'm hoping that #231 wouldn't need to the whole `enable CACHE_OPTIONS.SYMBOLIC_VALUES` thing. Looks like #231 has passed all CI, which feels promising.

The first step of this is to get number proxies to be plumbed through, I think that might be helpful for #231.
In terms of dynamic shape, we can slowly expanding its support.

Start off with allowing scalar input as number proxies, while still requiring tensor to be constant shape. I'm trying to figure out how/where to properly insert prologue_trace guard. Right now considering doing that from executor. i.e. nvfuser would require reduction dim(s) to be baked in as constant, while torchex doesn't care.

I'll try to come up with a design doc for review.

**Comment by jjsjann123** [comment]
A simple example:

```
def foo(a, scalar):
    scalar2 = scalar + scalar
    return (a + scalar2).sum(scalar)

import torch

dtype = torch.float32

size0 = 8
size1 = 16
size2 = 32
a = torch.randn(size0, size1, size2, device="cuda").to(dtype=dtype)

val=1

jfoo = thunder.jit(foo, cache="symbolic values")
out = jfoo(a, val)

```


```
@torch.no_grad()
@no_autocast
def prologue(*args, **kwargs):
  # args: "Any"
  prims.check_len(args, 2)
  # kwargs: "Any"
  prims.check_len(kwargs, 0)
  a: "cuda:0 f32[8, 16, 32]" = args[0]
  prims.assert_tensor_metadata_impl(a, (8, 16, 32), devices.Device("cuda:0"), dtypes.float32, False)
  scalar: "int 1" = args[1]
  prims.check_tensor_metadata(a, (8, 16, 32), 'cuda:0', torch.float32, False)
  cache_info: "Any" = thunder._get_cache_info()
  cache_info_is_autocast_enabled: "bool False" = cache_info['is_autocast_enabled']
  prims.check_number_type_and_value(cache_info_is_autocast_enabled, False)
  cache_info_no_grad_sync: "bool False" = cache_info['no_grad_sync']
  prims.check_number_type_and_value(cache_info_no_grad_sync, False)
  return (a, scalar)
```

compute trace
```
import thunder
import thunder.torch as ltorch
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a, scalar):
  # a: "cuda:0 f32[8, 16, 32]"
  # scalar: "int 1"
  result = ltorch.add(a, 2, alpha=None)  # result: "cuda:0 f32[8, 16, 32]"
    # _ = prims.convert_element_type(2, float)
    # result = prims.add(a, 2.0)  # result: "cuda:0 f32[8, 16, 32]"
  t2 = ltorch.sum(result, scalar, False, dtype=None)  # t2: "cuda:0 f32[8, 32]"
    # t2 = prims.sum(result, (scalar,))  # t2: "cuda:0 f32[8, 32]"
  return t2
```

Note that this is not a working example yet, i.e. nvfuser executor would bake in `scalar` as constant, even though it shows up as a symbolic value in the trace.

**Comment by jjsjann123** [comment]
Note this one is still on-going.

**Comment by jjsjann123** [comment]
archiving this PR since I'm occupied at this time. I'll come back to this afterwards.
